src/cluster/agglomerative.rs

```diff
@@ -0,0 +1,337 @@
+//!     AgglomerativeClustering()
+//!     >>> clustering.labels_
+//!     array([1, 1, 1, 0, 0, 0])
+//!
+//! ## Example:
+//!
+//! ```
+//! use smartcore::linalg::naive::dense_matrix::*;
+//! use smartcore::math::num::RealNumber;
+//! use smartcore::cluster::agglomerative::*;
+//! use smartcore::algorithm::neighbour::fastpair::FastPair;
+//!
+//! // Iris data
+//! let x = DenseMatrix::from_2d_array(&[
+//!            &[5.1, 3.5, 1.4, 0.2],
+//!            &[4.9, 3.0, 1.4, 0.2],
+//!            &[4.7, 3.2, 1.3, 0.2],
+//!            &[4.6, 3.1, 1.5, 0.2],
+//!            &[5.0, 3.6, 1.4, 0.2],
+//!            &[5.4, 3.9, 1.7, 0.4],
+//!            &[4.6, 3.4, 1.4, 0.3],
+//!            &[5.0, 3.4, 1.5, 0.2],
+//!            &[4.4, 2.9, 1.4, 0.2],
+//!            &[4.9, 3.1, 1.5, 0.1],
+//!            &[7.0, 3.2, 4.7, 1.4],
+//!            &[6.4, 3.2, 4.5, 1.5],
+//!            &[6.9, 3.1, 4.9, 1.5],
+//!            &[5.5, 2.3, 4.0, 1.3],
+//!            &[6.5, 2.8, 4.6, 1.5],
+//!            &[5.7, 2.8, 4.5, 1.3],
+//!            &[6.3, 3.3, 4.7, 1.6],
+//!            &[4.9, 2.4, 3.3, 1.0],
+//!            &[6.6, 2.9, 4.6, 1.3],
+//!            &[5.2, 2.7, 3.9, 1.4],
+//!            ]);
+//!
+//! // Fit to data, with a threshold
+//! // example using FastPair
+//! let cluster = AggregativeFastPair::fit(&x, 4).unwrap();
+//! // return results/labels/dendrogram
+//! let labels = cluster.labels();
+//! ```
+//!
+use crate::algorithm::neighbour::fastpair::FastPair;
+use crate::error::Failed;
+use crate::linalg::{BaseMatrix, Matrix};
+use crate::math::distance::euclidian::Euclidian;
+use crate::math::num::RealNumber;
+
+///
+/// Abstract trait for sequential, agglomerative, hierarchic, non-overlapping methods
+///
+pub trait SAHNClustering<T: RealNumber, M: Matrix<T>> {
+    ///
+    /// Aggregate the data according to given distance threshold
+    ///
+    fn fit(data: &M, k: usize) -> Result<AggregativeFastPair<T, M>, Failed>;
+
+    ///
+    /// Return clusters labels
+    ///
+    fn labels(&self) -> &M;
+
+    ///
+    /// Recompute labels according to a threshold
+    ///
+    fn labels_with_threhshold(&self, threshold: T);
+}
+
+///
+/// An implementation of Bottom-Up (Agglomerative) Hierarchical
+///  Clustering with `FastPair`
+///
+pub struct AggregativeFastPair<T: RealNumber, M: Matrix<T>> {
+    labels: Box<M>, // copmuted cluster labels
+    current: Option<T>,
+}
+
+// Add linkage algorithms
+impl<T: RealNumber, M: Matrix<T>> FastCluster<T> for AggregativeFastPair<T, M> {}
+
+// Implement aggregative clustering using FastPair and fastcluster
+impl<T: RealNumber, M: Matrix<T>> SAHNClustering<T, M> for AggregativeFastPair<T, M> {
+    //
+    // 1. Compute `FastPair` on matrix's rows
+    // 2. Port dissimilarities into upper-trinagular matrix
+    // 3. Compute dendrogram
+    //
+    // The linkage distance threshold above which clusters will not be merged.
+    fn fit(data: &M, _k: usize) -> Result<AggregativeFastPair<T, M>, Failed> {
```

Comment:
I don't think we need parameter k here. There should be 2 separate functions that split data by number of clusters, `k` and split data by distance threshold, `d`. Method fit should produce dendrogram and all necessary byproducts   that are required to make these splits.
What about linkage methods? How can I choose between “ward”, “complete”, “average”, “single” linkage types? I think linkage should be a parameter of fit.